src/state/apis/covalent/validators.ts

```diff
@@ -0,0 +1,241 @@
+  // [SupportedCovalentNetwork.BinanceSmartChain]: bscChainId,
+  // [SupportedCovalentNetwork.AvalancheCChain]: avalancheCChainId,
+} as const
+
+export const CHAIN_ID_TO_COVALENT_NETWORK_MAP = invert(
+  COVALENT_NETWORKS_TO_CHAIN_ID_MAP,
+) as Partial<Record<ChainId, SupportedCovalentNetwork>>
+export const covalentNetworkToChainId = (network: SupportedCovalentNetwork): ChainId | undefined =>
+  COVALENT_NETWORKS_TO_CHAIN_ID_MAP[network]
+
+export const chainIdToCovalentNetwork = (chainId: ChainId): SupportedCovalentNetwork | undefined =>
+  CHAIN_ID_TO_COVALENT_NETWORK_MAP[chainId]
+
+const NftItemSchema = z.object({
+  token_id: z.string(),
+  token_balance: z.string().nullable(),
+  token_url: z.string().nullable(),
+  supports_erc: z.array(z.string()).nullable(),
+  token_price_wei: z.string().nullable(),
+  token_quote_rate_eth: z.string().nullable(),
+  original_owner: z.string().nullable(),
+  external_data: z
+    .object({
+      name: z.string().nullable(),
+      description: z.string().nullable(),
+      image: z.string().nullable(),
+      image_256: z.string().nullable(),
+      image_512: z.string().nullable(),
+      image_1024: z.string().nullable(),
+      animation_url: z.string().nullable(),
+      external_url: z.string().nullable(),
+      attributes: z.array(z.object({ trait_type: z.string(), value: z.string() })).nullable(),
+      owner: z.string().nullable(),
+    })
+    .nullable(),
+  owner: z.string().nullable(),
+  owner_address: z.string().nullable(),
+  burned: z.boolean().nullable(),
+})
+
+const CovalentNftItemSchema = z.object({
+  contract_decimals: z.number().nullable(),
+  contract_name: z.string(),
+  contract_ticker_symbol: z.string(),
+  contract_address: z.string(),
+  supports_erc: z.array(z.string()).nullable(),
+  logo_url: z.string().nullable(),
+  last_transferred_at: z.string(),
+  native_token: z.boolean().nullable(),
+  type: z.string(),
+  balance: z.string(),
+  balance_24h: z.string(),
+  quote_rate: z.string().nullable(),
+  quote_rate_24h: z.string().nullable(),
+  quote: z.string().nullable(),
+  pretty_quote: z.string().nullable(),
+  quote_24h: z.string().nullable(),
+  pretty_quote_24h: z.string().nullable(),
+  nft_data: z.array(NftItemSchema).nullable(),
+  is_spam: z.boolean().nullable(),
+})
+
+const CovalentNftUserTokensResponseSchema = z.object({
+  data: z.object({
+    address: z.string(),
+    updated_at: z.string(),
+    next_update_at: z.string(),
+    quote_currency: z.string(),
+    chain_id: z.number(),
+    chain_name: z.string(),
+    items: z.array(CovalentNftItemSchema),
+    is_spam: z.boolean().nullable(),
+  }),
+  error: z.boolean(),
+  error_message: z.string().nullable(),
+  error_code: z.number().nullable(),
+})
+
+export type CovalentNftUserTokensResponseType = Infer<typeof CovalentNftUserTokensResponseSchema>
+export type CovalentNftItemSchemaType = Infer<typeof CovalentNftItemSchema>
+
+export const parseToV2NftUserItem = (
```

Comment:
Is there any reason we don't handle multiple NFTs on the same contract?

`nft_data` can contain one or more NFTs from the same collection (`contract`), but this logic just takes, and displays, the first in the array in the response data.

To test this, I dropped a few cents on junk NFTs from Planet IX, and only the first in the list shows when running this PR:

Opensea:

<img width="354" alt="Screenshot 2023-05-19 at 1 11 09 pm" src="https://github.com/shapeshift/web/assets/97164662/41a0a006-abba-4a5e-849a-457ed5189e60">

This PR:
<img width="730" alt="Screenshot 2023-05-19 at 1 12 23 pm" src="https://github.com/shapeshift/web/assets/97164662/b3d77917-b109-4892-a2a3-4e18b30491f8">

Here is a request with the response payload I'm referring to: https://api.covalenthq.com/v1/matic-mainnet/address/0xc2090e54b0db09a1515f203aea6ed62a115548ec/balances_v2/?key=ckey_2160df0b81ac46e9a642b35a5f1&nft=true

We also don't seem to have a way to show the quantity of NFTs, though covalent doesn't actually seem to return that data to us.

Reply:
gm 67ae3758e54a75fd2f81c4c720b357accada866d

<img width="1310" alt="image" src="https://github.com/shapeshift/web/assets/17035424/c3ddec3c-9b57-4a94-871e-02f73047fb28">

WRT the quantity, that's definitely a nice addition, though we don't aggregate by collection so we would have to decide with @shapeshift/product how to handle this. I'm going to have a Friday/week-end spike on how we represent/store NFT assets at a lower-level (i.e as regular CAIPs in the asset slice, in a diff slice altogether, as augmented `Asset`s or possibly another option I didn't yet come up with) which will make this possible from a technical perspective.

Currently the NFT fetching happens in its own, independent API with no associated store, which was the intended behavior but makes it harder to do things like aggregation since it's dumb and partially normalized only (i.e no `ids` or `collections` available)